Add guide about host binaries (#16198)
docker/docs: desktop/extensions-sdk/guides/invoke-host-binaries.md

---
title: Invoke host binaries
description: Add invocations to host binaries from the frontend with the extension SDK.
keywords: Docker, extensions, sdk, build
---

In some cases, your extension needs to invoke some command from the host (the computer of your users). For example, you 
might wand to invoke the CLI of your cloud provider to create a new resource, or the CLI of a tool your extension 
provides. 
You could do that executing the CLI from a container with the extension SDK. But this CLI needs to access the host's 
filesystem, which isn't possible if it runs in a container.

> Host binaries allows exactly this: to invoke binaries from the host. As extensions can run on multiple platforms, 

Reviewer: ```suggestion
Host binaries allow exactly this: to invoke from the extension those binaries that are shipped as part of your extension and have been deployed to the host. As extensions can run on multiple platforms, 
```

this means that you need to ship the binaries for all the platforms you want to support.

> 

Reviewer: Add a "info" msg clarifying that you cannot run any binary on the host, but only those that are shipped as part of the extension.

In this example, this CLI will be a simple `Hello world` script that must be invoked with a parameter and will return a 
string. Since the extension will support multiple platforms, the script will be written in `bash` for macOS and Linux, 
and in `powershell` for Windows.

> ## Add the binaries to the extension

Reviewer: I think this section is confusing as it mentions binaries but the examples are about shell scripts. I was expecting an example with a real binary, like `kubectl`.

Author: That is interesting! You are right. But, since any script can be used here, be it binaries or shell scripts, should we rename host binaries as host as scripts instead? 
I think I can reuse `kubectl` here but IIRC I thought it would be repetitive with the guide about interacting with k8s.

Author: I discussed with @gtardif we decided to talk about executables rather than binaries in the doc.

Reviewer: @benja-M-1 Executables seems to be a term more accurate. I hope we can say in the docs that we use the term binary and executable interchangeably (to not confuse users with the word `binaries` in `metadata.json`)

Create a file `binaries/unix/hello.sh` with the following content:

```bash
#!/bin/sh
echo "Hello, $1!"
```

Create another file `binaries/windows/hello.ps1` with the following content:

```powershell
Write-Output "Hello, $args[0]!"
```

Make them both executable:

```bash

> chmod +x binaries/*

Reviewer: If I'm following this tutorial on Windows, `chmod` won't work

Author: Indeed, I need to figure out if it is needed and how to make it work. I haven't tested it yet on windows. I wanted the team's feedback before pursuing it. Thanks for raising this though. Good reminder!

Author: Indeed, I need to figure out if it is needed and how to make it work. I haven't tested it yet on windows. I wanted the team's feedback before pursuing it. Thanks for raising this though. Good reminder!

```

> Then update the `Dockerfile` to copy the `binaries` folder into the extension container.

Reviewer: ```suggestion
Then update the `Dockerfile` to copy the `binaries` folder into the extension's image filesystem.
```

Author: According to [the docs](https://docs.docker.com/engine/reference/builder/#copy), it is copied to the container's filesystem rather than the image's one. I'll go with `container's filesystem` instead.

`The COPY instruction copies new files or directories from <src> and adds them to the filesystem of the container at the path <dest>.`

```dockerfile
# Copy the binaries into the right folder
COPY binaries/windows/hello.ps1 /windows/hello.ps1
COPY binaries/unix/hello.sh /linux/hello.sh
COPY binaries/unix/hello.sh /darwin/hello.sh
```

## Invoke the host binary from the UI

> In your app, use the Docker Desktop Client object and then invoke the binary provided by the extension with the 

Reviewer: ```suggestion
In your extension, use the Docker Desktop Client object to invoke the binary provided by the extension with the 
```

> `ddClient.extension.host.cli.exec()` function.

Reviewer: Add link to the API reference page

> In this example, the binary returns a string as result, obtained by `result?.stdout`, as soon as the app starts.

Reviewer: ```suggestion
In this example, the binary returns a string as result, obtained by `result?.stdout`, as soon as the extension view is rendered.
```

<ul class="nav nav-tabs">
  <li class="active"><a data-toggle="tab" data-target="#react-app" data-group="react">For React</a></li>
  <li><a data-toggle="tab" data-target="#vue-app" data-group="vue">For Vue</a></li>
  <li><a data-toggle="tab" data-target="#angular-app" data-group="angular">For Angular</a></li>
  <li><a data-toggle="tab" data-target="#svelte-app" data-group="svelte">For Svelte</a></li>
</ul>

<div class="tab-content">
  <div id="react-dockerfile" class="tab-pane fade in active" markdown="1">

```typescript

> 

Reviewer: ```suggestion
```

export function App() {
  const ddClient = createDockerDesktopClient();
  const [hello, setHello] = useState("");

  useEffect(() => {
    const run = async () => {
      let binary = "hello.sh";
      if (ddClient.host.platform === 'win32') {
        binary = "hello.ps1";
      }

      const result = await ddClient.extension.host?.cli.exec(binary, ["world"]);
      setHello(result?.stdout);

    };
    run();
  }, [ddClient]);
    
  return (
    <div>
      {hello}
    </div>
  );
}
```

  </div>
  <div id="vue-app" class="tab-pane fade" markdown="1">

<br/>

> **Important**

> > We don't have an example for Vue yet. [Fill out the form](https://docs.google.com/forms/d/e/1FAIpQLSdxJDGFJl5oJ06rG7uqtw1rsSBZpUhv_s9HHtw80cytkh2X-Q/viewform?usp=pp_url&entry.1333218187=Vue)

Reviewer: This painted door is great! 💯

> and let us know you'd like a sample with Vue.
{: .important }

  </div>
  <div id="angular-app" class="tab-pane fade" markdown="1">

<br/>

> **Important**
> We don't have an example for Angular yet. [Fill out the form](https://docs.google.com/forms/d/e/1FAIpQLSdxJDGFJl5oJ06rG7uqtw1rsSBZpUhv_s9HHtw80cytkh2X-Q/viewform?usp=pp_url&entry.1333218187=Angular)
> and let us know you'd like a sample with Angular.
{: .important }

  </div>
  <div id="svelte-app" class="tab-pane fade" markdown="1">

<br/>

> **Important**
> We don't have an example for Svelte yet. [Fill out the form](https://docs.google.com/forms/d/e/1FAIpQLSdxJDGFJl5oJ06rG7uqtw1rsSBZpUhv_s9HHtw80cytkh2X-Q/viewform?usp=pp_url&entry.1333218187=Svelte)
> and let us know you'd like a sample with Svelte.
{: .important }

  </div>
</div>

## Configure the metadata file

> The host binaries must be specified in the `metadata.json` so that Docker Desktop copies them on the host when installing the extension.

Reviewer: ```suggestion
The host binaries must be specified in the `metadata.json` so that Docker Desktop copies them on the host when installing the extension. Once the extension is uninstalled, the binaries that were copied will be removed as well.
```

```json
{
  "vm": {
    ...
  },
  "ui": {
    ...
  },
  "host": {
    "binaries": [
      {
        "darwin": [
          {
            "path": "/darwin/hello.sh"
          }
        ],
        "linux": [
          {
            "path": "/linux/hello.sh"
          }
        ],
        "windows": [
          {
            "path": "/windows/hello.ps1"
          }
        ]
      }
    ]
  }
}
```

> The `path` must reference the path of the binary inside the container.

Reviewer: ```suggestion
The `path` must reference the path of the binary inside the extension's image filesystem.
```
